src/transport_ofi.c

```diff
@@ -312,179 +313,224 @@ typedef enum{
+void shmem_transport_ofi_stx_allocate(shmem_transport_ctx_t *ctx)
 {
-    /* STX contexts are shared by SHMEM contexts that are private to the same
-     * thread (i.e. have SHMEM_CTX_PRIVATE option set and same thread ID).  */
+    /* SHMEM contexts that are private to the same thread (i.e. have
+     * SHMEM_CTX_PRIVATE option set) share the same STX.  */
     if (ctx->options & SHMEM_CTX_PRIVATE) {
 
         shmem_transport_ofi_stx_kvs_t *f;
         HASH_FIND_INT(shmem_transport_ofi_stx_kvs, &ctx->tid, f);
+
         if (f) {
-            /* A thread-private STX already exists for this thread id, so increment the
-             * reference count and rollback to the last index into the STX resource pool:  */
-            f->stx->ref_cnt++;
+            shmem_transport_ofi_stx_pool[f->stx_idx].ref_cnt++;
             ctx->stx_idx = f->stx_idx;
 
         } else {
-            shmem_transport_ofi_stx_pool_search(ctx);
+            /* No STX allocated to the given TID, attempt to allocate one */
+            int is_unused = 1;
+            int stx_idx;
+            shmem_transport_ofi_stx_t *stx = NULL;
+
+            stx_idx = shmem_transport_ofi_stx_search_unused();
+
+            /* Couldn't get new STX, assign a shared one */
+            /* Note: shared STX allocation is always successful */
+            if (stx_idx < 0) {
+                DEBUG_STR("private STX unavailable, falling back to STX sharing");
+                is_unused = 0;
+                stx_idx = shmem_transport_ofi_stx_search_shared();
+            }
+
+            shmem_internal_assert(stx_idx >= 0);
+            stx = &shmem_transport_ofi_stx_pool[stx_idx];
+            ctx->stx_idx = stx_idx;
+            stx->ref_cnt++;
+
             shmem_transport_ofi_stx_kvs_t *e = malloc(sizeof(shmem_transport_ofi_stx_kvs_t));
             if (e == NULL) {
-                RAISE_ERROR_STR("Error: out of memory when allocating OFI STX key/value");
+                RAISE_ERROR_STR("out of memory when allocating STX KVS entry");
             }
-            e->tid = ctx->tid;
+            e->tid     = ctx->tid;
             e->stx_idx = ctx->stx_idx;
-            e->stx = &shmem_transport_ofi_stx_pool[ctx->stx_idx];
-            e->stx->ref_cnt++;
-            e->stx->is_private = true;
 
-            HASH_ADD_INT(shmem_transport_ofi_stx_kvs, tid, e);
+            if (is_unused) {
+                stx->is_private = 1;
+                HASH_ADD_INT(shmem_transport_ofi_stx_kvs, tid, e);
+            } else {
+                ctx->options &= ~SHMEM_CTX_PRIVATE;
+            }
         }
+    /* TODO: Optimize this case? else if (ctx->options & SHMEM_CTX_PRIVATE) */
+    } else {
+        int stx_idx = shmem_transport_ofi_stx_search_unused();
 
+        /* TODO: Currently only support greedy assignment of STX to non-private
+         * contexts.  In the future, we may want to set a threshold for refs on
+         * a shared STX before allocating the next one. */
```

Comment:
👍   So we'd have a threshold in `search_unused` that applies only shared contexts?